### ISSUE TITLE
Tectonic Console does not have stable connection to APIServer causing its data to not load

### ISSUE THREAD

**Issue by metral** [coreos/tectonic-installer#954]
## What keywords did you search in tectonic-installer issues before filing this one?
tectonic console apiserver

## Is this a BUG REPORT or FEATURE REQUEST?

bug report

## Versions

* Tectonic version ([release](https://github.com/coreos/tectonic-installer/releases) or commit hash): 1.6.4-tectonic.1
* Terraform version (`terraform version`): Terraform v0.9.6-dev (fcdf49)
* Platform (aws|azure|openstack|metal): Azure using all existing networking components (NSG, private vnet/subnets, DNS) that currently live in a fork.

## What happened?

After creating a cluster on Azure that uses existing networking components, the Tectonic console returns errors regarding communication with the APIServer. This causes the console to not load any data and return errors as found in the images below. This lasts for about 10 minutes before working again as expected.

Though changes were made to 1.6.4-tectonic.1 to support existing networking components on Azure, a [similar issue ](https://github.com/coreos/tectonic-forum/issues/132)has been filed in the forums related to the console not loading data and the `http: proxy error: context canceled` error shown in the tectonic console logs below.

#### Console not working on boot shown on home and other resource pages:
![image](https://cloud.githubusercontent.com/assets/1112768/26698887/8d2458b8-46cc-11e7-905d-3443bcf04afa.png)

![image](https://cloud.githubusercontent.com/assets/1112768/26698928/c1c93a66-46cc-11e7-84d3-57d313fe845a.png)

![image](https://cloud.githubusercontent.com/assets/1112768/26698936/ca7c0f76-46cc-11e7-9ab2-2b2ad53904e0.png)

#### And after about ~10 minutes of cluster creation, it all just starts working again.

I've encountered this on a couple of runs now:

![image](https://cloud.githubusercontent.com/assets/1112768/26698988/f57b43e0-46cc-11e7-9bac-a4066382ede7.png)

![image](https://cloud.githubusercontent.com/assets/1112768/26699252/f35d6754-46cd-11e7-89e6-ffbc79c51551.png)

#### Tectonic Console Logs:

```
$ kubectl logs tectonic-console-962128392-0nqb5 -n tectonic-system

2017/06/1 20:08:34 http: Provider config sync failed, retrying in 1s: Get https://metral041.example.com/identity/.well-known/openid-configuration: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
2017/06/1 20:08:40 http: Provider config sync still failing, retrying in 2s: Get https://metral041.example.com/identity/.well-known/openid-configuration: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
2017/06/1 20:08:42 http: Updating provider config: config=oidc.ProviderConfig{Issuer:(*url.URL)(0xc420222380), AuthEndpoint:(*url.URL)(0xc420222400), TokenEndpoint:(*url.URL)(0xc420222480), UserInfoEndpoint:(*url.URL)(nil), KeysEndpoint:(*url.URL)(0xc420222500), RegistrationEndpoint:(*url.URL)(nil), ScopesSupported:[]string{"openid", "email", "groups", "profile", "offline_access"}, ResponseTypesSupported:[]string{"code"}, ResponseModesSupported:[]string(nil), GrantTypesSupported:[]string(nil), ACRValuesSupported:[]string(nil), SubjectTypesSupported:[]string{"public"}, IDTokenSigningAlgValues:[]string{"RS256"}, IDTokenEncryptionAlgValues:[]string(nil), IDTokenEncryptionEncValues:[]string(nil), UserInfoSigningAlgValues:[]string(nil), UserInfoEncryptionAlgValues:[]string(nil), UserInfoEncryptionEncValues:[]string(nil), ReqObjSigningAlgValues:[]string(nil), ReqObjEncryptionAlgValues:[]string(nil), ReqObjEncryptionEncValues:[]string(nil), TokenEndpointAuthMethodsSupported:[]string{"client_secret_basic"}, TokenEndpointAuthSigningAlgValuesSupported:[]string(nil), DisplayValuesSupported:[]string(nil), ClaimTypesSupported:[]string(nil), ClaimsSupported:[]string{"aud", "email", "email_verified", "exp", "iat", "iss", "locale", "name", "sub"}, ServiceDocs:(*url.URL)(nil), ClaimsLocalsSupported:[]string(nil), UILocalsSupported:[]string(nil), ClaimsParameterSupported:false, RequestParameterSupported:false, RequestURIParamaterSupported:false, RequireRequestURIRegistration:false, Policy:(*url.URL)(nil), TermsOfService:(*url.URL)(nil), ExpiresAt:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}
2017/06/1 20:08:42 http: Provider config sync no longer failing
2017/06/1 20:08:43 http: Updating provider config: config=oidc.ProviderConfig{Issuer:(*url.URL)(0xc420225b80), AuthEndpoint:(*url.URL)(0xc420225c00), TokenEndpoint:(*url.URL)(0xc420225c80), UserInfoEndpoint:(*url.URL)(nil), KeysEndpoint:(*url.URL)(0xc420225d00), RegistrationEndpoint:(*url.URL)(nil), ScopesSupported:[]string{"openid", "email", "groups", "profile", "offline_access"}, ResponseTypesSupported:[]string{"code"}, ResponseModesSupported:[]string(nil), GrantTypesSupported:[]string(nil), ACRValuesSupported:[]string(nil), SubjectTypesSupported:[]string{"public"}, IDTokenSigningAlgValues:[]string{"RS256"}, IDTokenEncryptionAlgValues:[]string(nil), IDTokenEncryptionEncValues:[]string(nil), UserInfoSigningAlgValues:[]string(nil), UserInfoEncryptionAlgValues:[]string(nil), UserInfoEncryptionEncValues:[]string(nil), ReqObjSigningAlgValues:[]string(nil), ReqObjEncryptionAlgValues:[]string(nil), ReqObjEncryptionEncValues:[]string(nil), TokenEndpointAuthMethodsSupported:[]string{"client_secret_basic"}, TokenEndpointAuthSigningAlgValuesSupported:[]string(nil), DisplayValuesSupported:[]string(nil), ClaimTypesSupported:[]string(nil), ClaimsSupported:[]string{"aud", "email", "email_verified", "exp", "iat", "iss", "locale", "name", "sub"}, ServiceDocs:(*url.URL)(nil), ClaimsLocalsSupported:[]string(nil), UILocalsSupported:[]string(nil), ClaimsParameterSupported:false, RequestParameterSupported:false, RequestURIParamaterSupported:false, RequireRequestURIRegistration:false, Policy:(*url.URL)(nil), TermsOfService:(*url.URL)(nil), ExpiresAt:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}
2017/06/1 20:08:43 cmd/main: Binding to 0.0.0.0:8080...
2017/06/1 20:08:43 cmd/main: not using TLS
2017/06/1 20:09:23 server: no token found: token cookie not found in request: http: named cookie not present
2017/06/1 20:09:23 server: no token found: token cookie not found in request: http: named cookie not present
2017/06/1 20:09:23 server: no token found: token cookie not found in request: http: named cookie not present
2017/06/1 20:09:23 server: no token found: token cookie not found in request: http: named cookie not present
2017/06/1 20:09:23 server: no token found: token cookie not found in request: http: named cookie not present
2017/06/1 20:10:46 http: proxy error: EOF
2017/06/1 20:11:15 http: proxy error: context canceled
2017/06/1 20:11:46 http: proxy error: context canceled
2017/06/1 20:11:46 http: proxy error: context canceled
2017/06/1 20:12:15 http: proxy error: context canceled
2017/06/1 20:12:15 http: proxy error: context canceled
2017/06/1 20:12:15 http: proxy error: context canceled
2017/06/1 20:12:52 http: proxy error: context canceled
2017/06/1 20:12:52 http: proxy error: context canceled
2017/06/1 20:12:52 http: proxy error: context canceled
2017/06/1 20:13:25 http: proxy error: context canceled
2017/06/1 20:13:25 http: proxy error: context canceled
2017/06/1 20:13:25 http: proxy error: context canceled
2017/06/1 20:14:26 http: proxy error: context canceled
2017/06/1 20:14:26 http: proxy error: context canceled
2017/06/1 20:14:26 http: proxy error: context canceled
2017/06/1 20:15:26 http: proxy error: context canceled
2017/06/1 20:15:26 http: proxy error: context canceled
2017/06/1 20:15:26 http: proxy error: context canceled
2017/06/1 20:16:27 http: proxy error: context canceled
2017/06/1 20:16:27 http: proxy error: context canceled
2017/06/1 20:16:27 http: proxy error: context canceled
2017/06/1 20:17:27 http: proxy error: context canceled
2017/06/1 20:17:27 http: proxy error: context canceled
2017/06/1 20:17:27 http: proxy error: context canceled
2017/06/1 20:17:27 http: proxy error: context canceled
2017/06/1 20:17:27 http: proxy error: context canceled
2017/06/1 20:18:28 http: proxy error: context canceled
```

#### API Server Logs:

The only info found in here are TLS endpoint and Resetting endpoings messages that also exist in other clusters, so not sure if they have any significance to the issue related to the Console

```
$ kubectl logs kube-apiserver-01dkm -n kube-system

I0601 20:27:25.001535       5 logs.go:41] http: TLS handshake error from <IP_ADDR>:49927: EOF
W0601 20:27:29.327907       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes 7d441ce3-4705-11e7-9677-000d3a316094 4515 0 2017-06-01 20:04:16.011388 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{10.18.140.20  <nil> <nil>}] [] [{https 443 TCP}]}]}
W0601 20:27:39.337480       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes 7d441ce3-4705-11e7-9677-000d3a316094 4536 0 2017-06-01 20:04:16.011388 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{10.18.140.20  <nil> <nil>}] [] [{https 443 TCP}]}]}
I0601 20:27:40.001784       5 logs.go:41] http: TLS handshake error from <IP_ADDR>:50040: EOF
W0601 20:27:49.346616       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes 7d441ce3-4705-11e7-9677-000d3a316094 4556 0 2017-06-01 20:04:16.011388 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{10.18.140.20  <nil> <nil>}] [] [{https 443 TCP}]}]}
I0601 20:27:55.003148       5 logs.go:41] http: TLS handshake error from <IP_ADDR>:50169: EOF
W0601 20:27:59.356433       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes 7d441ce3-4705-11e7-9677-000d3a316094 4577 0 2017-06-01 20:04:16.011388 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{10.18.140.20  <nil> <nil>}] [] [{https 443 TCP}]}]}
W0601 20:28:09.365852       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes 7d441ce3-4705-11e7-9677-000d3a316094 4598 0 2017-06-01 20:04:16.011388 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{10.18.140.20  <nil> <nil>}] [] [{https 443 TCP}]}]}
I0601 20:28:10.005603       5 logs.go:41] http: TLS handshake error from <IP_ADDR>:50258: EOF
```

## What you expected to happen?

For the console to work out of the box

## How to reproduce it (as minimally and precisely as possible)?
Given that I'm on a fork and working in a special environment, this is not easily reproducible. I'm hoping the images & logs help identify a possible root cause.

**Comment by metral** [comment]
Signing out and back in seems to alleviate this issue, but it still occurs.


Logs from tectonic-identity:

```
/kubectl logs tectonic-identity-3240565221-9k609 -n tectonic-system
time="2017-06-01T22:27:25Z" level=info msg="config issuer: https://metral045.example.com/identity"
time="2017-06-01T22:27:25Z" level=info msg="config connector: local passwords enabled"
time="2017-06-01T22:27:25Z" level=error msg="create third party resource auth-code.oidc.coreos.com"
time="2017-06-01T22:27:25Z" level=error msg="create third party resource auth-request.oidc.coreos.com"
time="2017-06-01T22:27:25Z" level=error msg="create third party resource o-auth2-client.oidc.coreos.com"
time="2017-06-01T22:27:25Z" level=error msg="create third party resource signing-key.oidc.coreos.com"
time="2017-06-01T22:27:25Z" level=error msg="create third party resource refresh-token.oidc.coreos.com"
time="2017-06-01T22:27:25Z" level=error msg="create third party resource password.oidc.coreos.com"
time="2017-06-01T22:27:25Z" level=error msg="create third party resource offline-sessions.oidc.coreos.com"
time="2017-06-01T22:27:25Z" level=error msg="create third party resource connector.oidc.coreos.com"
time="2017-06-01T22:27:25Z" level=info msg="config storage: kubernetes"
time="2017-06-01T22:27:25Z" level=info msg="config static client: tectonic-console"
time="2017-06-01T22:27:25Z" level=info msg="config static client: tectonic-kubectl"
time="2017-06-01T22:27:25Z" level=info msg="config skipping approval screen"
time="2017-06-01T22:27:25Z" level=info msg="keys expired, rotating"
time="2017-06-01T22:27:27Z" level=error msg="failed to rotate keys: not found"
time="2017-06-01T22:27:27Z" level=info msg="listening (http) on 0.0.0.0:5556"
time="2017-06-01T22:27:27Z" level=info msg="listening (grpc) on 0.0.0.0:5557"
time="2017-06-01T22:27:57Z" level=info msg="keys expired, rotating"
time="2017-06-01T22:28:01Z" level=info msg="keys rotated, next rotation: 2017-06-02 04:28:01.3288784 +0000 UTC"
time="2017-06-01T22:28:31Z" level=info msg="keys expired, rotating"
time="2017-06-01T22:28:35Z" level=info msg="keys rotated, next rotation: 0001-01-01 00:00:00 +0000 UTC"
time="2017-06-01T22:29:05Z" level=info msg="keys expired, rotating"
time="2017-06-01T22:29:09Z" level=info msg="keys rotated, next rotation: 0001-01-01 00:00:00 +0000 UTC"
time="2017-06-01T22:29:39Z" level=info msg="keys expired, rotating"
time="2017-06-01T22:29:47Z" level=info msg="keys rotated, next rotation: 0001-01-01 00:00:00 +0000 UTC"
time="2017-06-01T22:30:11Z" level=error msg="failed to create ID token: no key to sign payload with"
time="2017-06-01T22:46:19Z" level=error msg="Storage health check failed: create auth request: POST https://10.3.0.1:443/apis/oidc.coreos.com/v1/namespaces/tectonic-system/authrequests Forbidden: response from server \"User \"system:serviceaccount:tectonic-system:default\" cannot create authrequests.oidc.coreos.com in the namespace \"tectonic-system\".\""
time="2017-06-01T23:02:27Z" level=info msg="garbage collection run, delete auth requests=0, auth codes=2"
time="2017-06-02T09:52:06Z" level=error msg="Storage health check failed: create auth request: POST https://10.3.0.1:443/apis/oidc.coreos.com/v1/namespaces/tectonic-system/authrequests Internal Server Error: response from server \"{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"etcdserver: request timed out, possibly due to previous leader failure\",\"code\":500}\""
time="2017-06-02T10:30:21Z" level=info msg="keys expired, rotating"
time="2017-06-02T10:30:26Z" level=info msg="keys rotated, next rotation: 0001-01-01 00:00:00 +0000 UTC"
time="2017-06-02T10:30:56Z" level=info msg="keys expired, rotating"
time="2017-06-02T10:30:57Z" level=info msg="keys rotated, next rotation: 2017-06-02 16:30:57.3209042 +0000 UTC"
time="2017-06-02T10:31:27Z" level=info msg="keys expired, rotating"
time="2017-06-02T10:31:30Z" level=info msg="keys rotated, next rotation: 0001-01-01 00:00:00 +0000 UTC"
time="2017-06-02T10:32:00Z" level=info msg="keys expired, rotating"
time="2017-06-02T10:32:01Z" level=info msg="keys rotated, next rotation: 0001-01-01 00:00:00 +0000 UTC"
time="2017-06-02T10:32:31Z" level=info msg="keys expired, rotating"
time="2017-06-02T10:32:34Z" level=info msg="keys rotated, next rotation: 2017-06-02 16:32:34.2189057 +0000 UTC"
time="2017-06-02T10:33:04Z" level=info msg="keys expired, rotating"
time="2017-06-02T10:33:05Z" level=info msg="keys rotated, next rotation: 2017-06-02 16:33:05.2252651 +0000 UTC"
```

Recent logs for tectonic-console:

```
2017/06/2 05:37:57 http: proxy error: context canceled
2017/06/2 05:38:04 http: proxy error: context canceled
2017/06/2 05:38:47 http: proxy error: context canceled
2017/06/2 05:38:47 http: proxy error: context canceled
2017/06/2 05:38:59 server: no token found: token cookie not found in request: http: named cookie not present
2017/06/2 05:38:59 server: no token found: token cookie not found in request: http: named cookie not present
2017/06/2 05:38:59 server: no token found: token cookie not found in request: http: named cookie not present
2017/06/2 05:38:59 server: no token found: token cookie not found in request: http: named cookie not present
2017/06/2 05:38:59 server: no token found: token cookie not found in request: http: named cookie not present
2017/06/2 05:38:59 server: no token found: token cookie not found in request: http: named cookie not present
2017/06/2 05:39:09 Failed to dial backend: websocket.Dial wss://10.3.0.1:443/api/v1/namespaces?watch=true: bad status
2017/06/2 05:39:09 Failed to dial backend: websocket.Dial wss://10.3.0.1:443/api/v1/nodes?watch=true: bad status
2017/06/2 05:39:09 Failed to dial backend: websocket.Dial wss://10.3.0.1:443/api/v1/nodes?fieldSelector=metadata.name%3Dtectonic-worker-000000&watch=true: bad status
2017/06/2 05:39:13 Failed to dial backend: websocket.Dial wss://10.3.0.1:443/api/v1/nodes?watch=true: bad status
2017/06/2 05:39:59 httputil: ReverseProxy read error during body copy: context canceled
2017/06/2 05:40:01 Failed to dial backend: websocket.Dial wss://10.3.0.1:443/api/v1/secrets?watch=true: bad status
2017/06/2 05:40:01 Failed to dial backend: websocket.Dial wss://10.3.0.1:443/api/v1/events?watch=true: bad status
2017/06/2 05:40:01 httputil: ReverseProxy read error during body copy: context canceled
2017/06/2 05:40:14 httputil: ReverseProxy read error during body copy: context canceled
2017/06/2 05:40:25 httputil: ReverseProxy read error during body copy: context canceled
2017/06/2 05:43:55 httputil: ReverseProxy read error during body copy: context canceled
2017/06/2 15:49:14 server: error verifying token: oidc: unable to verify JWT signature: no matching keys
2017/06/2 15:49:14 server: error verifying token: oidc: unable to verify JWT signature: no matching keys
2017/06/2 15:49:14 server: error verifying token: oidc: unable to verify JWT signature: no matching keys
2017/06/2 15:49:14 server: error verifying token: oidc: unable to verify JWT signature: no matching keys
2017/06/2 15:49:14 server: error verifying token: oidc: unable to verify JWT signature: no matching keys
2017/06/2 15:49:14 server: error verifying token: oidc: unable to verify JWT signature: no matching keys
2017/06/2 15:49:14 server: no token found: token cookie not found in request: http: named cookie not present
2017/06/2 15:49:14 server: error verifying token: oidc: unable to verify JWT signature: no matching keys
2017/06/2 15:49:14 server: no token found: token cookie not found in request: http: named cookie not present
2017/06/2 15:49:21 auth: oauth success, redirecting to: "https://metral045.example.com/"
2017/06/2 15:49:31 http: proxy error: context canceled
2017/06/2 15:51:30 Failed to dial backend: websocket.Dial wss://10.3.0.1:443/api/v1/namespaces?watch=true: bad status
2017/06/2 15:53:43 Failed to dial backend: websocket.Dial wss://10.3.0.1:443/api/v1/namespaces?watch=true: bad status
2017/06/2 15:53:43 Failed to dial backend: websocket.Dial wss://10.3.0.1:443/api/v1/pods?watch=true: bad status
```

**Comment by metral** [comment]
The main difference my environment's have over default, is that in a private cluster setup in Azure, I must resort to standing up a nginx proxy to front the k8s API server as Internal Load Balancers don't support SNAT: https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-internal-overview#limitations, a feature bootkube and the self-hosted apiserver need to self-reference the LB they sit behind at the time of cluster bootstrapping.

The nginx proxy is configured with the stream module to proxy over all 443 to the ILB as a work around, and this functions fine, but could be the cause of the random TLS handshake errors being received by the api server, in addition to the standard ones from the DNS server as logs would typically show.

It may be that latency tied up with the nginx proxy could be causing enough of an interrupt to create random blips of unresponsiveness from the apiserver on services that communicate with it, but this is not fully certain yet.

**Comment by metral** [comment]
Stood up another cluster in the same env, and spotted these errors in the apiserver logs around the same time I experienced issues with the console not being able to communicate with the apiserver to pull data:

Note the error: `Unable to authenticate the request due to an error: [invalid bearer token, [crypto/rsa: verification error, oidc: unable to verify JWT signature: no matching keys]]`

```
I0606 21:37:47.382389       5 logs.go:41] http: TLS handshake error from <DNS_IP>:59403: EOF
I0606 21:37:53.407064       5 logs.go:41] http: TLS handshake error from <CUSTOM_PROXY_IP>:46992: EOF
W0606 21:37:55.706331       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes d056a9a7-4a58-11e7-ac3c-000d3a360680 162710 0 2017-06-06 01:38:17.1007385 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{<MASTER#2_IP>  <nil
> <nil>}] [] [{https 443 TCP}]}]}
I0606 21:38:02.384197       5 logs.go:41] http: TLS handshake error from <DNS_IP>:59538: EOF
W0606 21:38:05.729157       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes d056a9a7-4a58-11e7-ac3c-000d3a360680 162730 0 2017-06-06 01:38:17.1007385 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{<MASTER#2_IP>  <nil
> <nil>}] [] [{https 443 TCP}]}]}
I0606 21:38:08.407297       5 logs.go:41] http: TLS handshake error from <CUSTOM_PROXY_IP>:47028: EOF
W0606 21:38:15.754008       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes d056a9a7-4a58-11e7-ac3c-000d3a360680 162750 0 2017-06-06 01:38:17.1007385 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{<MASTER#2_IP>  <nil
> <nil>}] [] [{https 443 TCP}]}]}
I0606 21:38:17.385608       5 logs.go:41] http: TLS handshake error from <DNS_IP>:59660: EOF
I0606 21:38:23.407480       5 logs.go:41] http: TLS handshake error from <CUSTOM_PROXY_IP>:47070: EOF
W0606 21:38:25.774812       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes d056a9a7-4a58-11e7-ac3c-000d3a360680 162773 0 2017-06-06 01:38:17.1007385 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{<MASTER#2_IP>  <nil
> <nil>}] [] [{https 443 TCP}]}]}
E0606 21:38:27.623022       5 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [crypto/rsa: verification error, oidc: failed syncing KeySet: Get https://myprefix-metral074.cdx.vpc.starbucks.net/identity/keys: dial tcp <CONSOLE_IP>:443: getsockopt: connection timed out]
]
E0606 21:38:27.623125       5 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [crypto/rsa: verification error, oidc: unable to verify JWT signature: no matching keys]]
E0606 21:38:27.623022       5 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [crypto/rsa: verification error, oidc: unable to verify JWT signature: no matching keys]]
E0606 21:38:27.623177       5 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [crypto/rsa: verification error, oidc: unable to verify JWT signature: no matching keys]]
E0606 21:38:27.623231       5 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [crypto/rsa: verification error, oidc: unable to verify JWT signature: no matching keys]]
E0606 21:38:27.623278       5 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [crypto/rsa: verification error, oidc: unable to verify JWT signature: no matching keys]]
E0606 21:38:27.623354       5 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [crypto/rsa: verification error, oidc: unable to verify JWT signature: no matching keys]]
E0606 21:38:27.623362       5 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [crypto/rsa: verification error, oidc: unable to verify JWT signature: no matching keys]]
E0606 21:38:27.659417       5 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [crypto/rsa: verification error, oidc: unable to verify JWT signature: no matching keys]]
```

**Comment by sym3tri** [comment]
There are a number of issues (potentially unrelated) based on these logs. I'll comment below on each individually. Hopefully this is useful for this and future debugging.

---

**API Server**

```
W0606 21:37:55.706331       5 controller.go:386] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default /api/v1/namespaces/default/endpoints/kubernetes d056a9a7-4a58-11e7-ac3c-000d3a360680 162710 0 2017-06-06 01:38:17.1007385 +0000 UTC <nil> <nil> map[] map[] [] [] } [{[{<MASTER#2_IP>  <nil
> <nil>}] [] [{https 443 TCP}]}]}
```
This is a known issue and is unfortunately the "normal" behavior for now until this gets fixed upstream. Likely a red herring. For context see: https://github.com/coreos/tectonic-installer/pull/136

```
E0606 21:38:27.623125       5 authentication.go:58] Unable to authenticate the request due to an error: [invalid bearer token, [crypto/rsa: verification error, oidc: unable to verify JWT signature: no matching keys]]
```

This means the API server cannot reach Dex and can't sync the keys from Dex. Verify Dex is up and running without problems and can be reached at the configured URL from within the cluster.


---

**Console**
```
http: proxy error: context canceled
```

This likely means Console requests to the API server are taking longer than 30 seconds and timing out. The Console can't reach the API server or the API server is unresponsive. So this is likely a network issue. I would suspect your DNS, Load Balancer, or Nginx. 

For the next step in debugging I'd try to reproduce the error by hitting the API server directly at Console's configured URL (https://github.com/coreos/tectonic-installer/blob/master/modules/tectonic/resources/manifests/console/deployment.yaml#L40). See if you can reproduce the error with kubectl or curl. If you can, Console is not the culprit (I really doubt it is).

```
server: no token found: token cookie not found in request: http: named cookie not present
```
This means exactly what it says. No cookie with an auth token was set in your request. This appears normally when Console is first loaded in your browser and you haven't logged in yet. This only indicates 
a problem if you see it repeatedly _after_ you've already logged in.

```
server: error verifying token: oidc: unable to verify JWT signature: no matching keys
```

Same as the API Server log. Console can't reach Dex at the configured URL (https://github.com/coreos/tectonic-installer/blob/master/modules/tectonic/resources/manifests/console/deployment.yaml#L54-L58)

---

NOTE: the issuer URL (where Dex is running) must be configured exactly the same in: Dex, Console, and the API server.

TL;DR
All of these logs point to some sort of flakiness with the underlying network setup (i.e. load balancers sending traffic to nodes with no Dex or API server running, or the nginx being flaky). So I would manually test that each of those are performing as expected.

**Comment by metral** [comment]
The fix for the Tectonic Console & Identity issues seen were fixed by placing an nginx proxy in front of the Console LB.

The fix is needed because Azure Internal Load Balancer's don't support SNAT: https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-internal-overview#limitations

SNAT is required because the Console LB is configured with the masters as its address pool, and not having SNAT creates issues when the apiserver tries to communicate with the Console LB it sits behind.

**Comment by nottinhill** [comment]
I am seeing the same behaviour in AWS, and also username and password is not accepted.
